src/Microsoft.AspNet.Mvc.Core/DefaultControllerTypeProvider.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Reflection;
+using Microsoft.AspNet.Mvc.Core;
+using Microsoft.AspNet.Mvc.Logging;
+using Microsoft.Framework.Logging;
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// A <see cref="IControllerTypeProvider"/> that identifies controller types from assemblies
+    /// specified by the registered <see cref="IAssemblyProvider"/>.
+    /// </summary>
+    public class DefaultControllerTypeProvider : IControllerTypeProvider
+    {
+        private readonly IAssemblyProvider _assemblyProvider;
+        private readonly ILogger _logger;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="DefaultControllerTypeProvider"/>.
+        /// </summary>
+        /// <param name="assemblyProvider"><see cref="IAssemblyProvider"/> that provides assemblies to look for
+        /// controllers in.</param>
+        /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
+        public DefaultControllerTypeProvider(IAssemblyProvider assemblyProvider,
```

Comment:
For UseControllersFromServiceProvider with assemblies, call into this with a `StaticAssemblyListProvider`. Additionally move the heurisitics code into here.